fix: Portal 의존성이 없는 ToastCore 분리 및 Toast Portal 생성

iyu88 opened the pull request:
## 🤠 개요

- #89 

## 💫 설명

- Toast의 스토리를 수정할 때 확인한 사항입니다. 

- 기본적으로 Portal 로 감싸져 있기 때문에 Portal의 default id인 'cds-portal-root' 에 Toast가 렌더링됩니다. 

- 이렇게 하면 Storybook Docs에서 확인할 수 있듯이 여러 개의 Toast가 하나의 페이지에 존재할 경우 **맨 처음에 있는 'cds-portal-root' 에 모든 Toast가 렌더링됩니다.** 

- Toast를 감싸고 있는 Portal을 제거하고 앞으로 사용할 때는 명시적으로 Portal 로 감싸주는게 좋을 것 같습니다. 

```
<Portal>
  <Toast {...args} />
</Portal>
```

- Storybook Docs를 수정하는 PR이랑 합쳐지면 스토리가 정상적으로 표시될 것 같습니다!!

iyu88 committed:
fix: Toast 컴포넌트에서 Portal 제거

Toast 컴포넌트에서 Portal을 제거합니다.
Toast와 Portal이 강결합되어 있는 구조는 Story에서 개별적인 Toast UI를 확인하기 어렵게 만듭니다.
가령 기존의 구조에서는 맨 처음에 생기는 Portal Id인 'cds-portal-root' 에 모든 Toast가 렌더링됩니다.
따라서 Portal을 분리하고 사용하는 곳에서 명시적으로 Portal 컴포넌트를 사용하도록 합니다.

se030 reviewed (commented):
정리가 덜 된 상태에서 댓글을 썼다가 설명을 이상하게 했다는걸 깨달았어요. ㅠㅠ 이전 댓글을 읽으셨다면 무시해주세요.

1. 토스트들이 첫번째 스토리에 전부 나타나는건 똑같은 노드를 포털로 사용하기 때문인데, 이건 어플리케이션에서도 아래처럼 모든 토스트 메세지가 같은 영역을 사용하는게 자연스럽다고 생각해요.
  ![image](https://user-images.githubusercontent.com/63814960/229297404-332ec65d-7324-431e-aaad-60c479b9c0a0.png)

2. 현재 설계대로라면 Toast에서 Portal을 제거해도 스토리마다 Portal을 감싸면 같은 상황이 되지않나요? 저는 사용할 때마다 Portal을 붙여야 하는건 좋은 시그널이 아닌 것 같아요.

3. 대신 Docs 탭에서 발생한 이슈의 원인은 Portal CSS라는 생각이 들어요.
  - CdsProvider가 데코레이터를 통해서 스토리마다 감싸지기 때문에 id가 중복되는 포털 용 div가 여러개 생긴다는 문제가 있지만, 첫번째 노드만 사용하게 돼서 저희가 의도하는 동작과 달라지지는 않아요.
  - 다만 모든 토스트들이 첫번째 스토리 영역 안에 갇히는데, 포털을 통해 띄워지는 토스트들의 위치를 뷰포트 기준으로 잡아주면 될 것 같아요.
  - 이런 방식으로 여러개의 토스트를 고려하려면 위치값을 고려해야해서 추가 작업이 필요할지도 모르겠네요 ㅠㅠ 사용하기에 더 편한 방식으로 리팩토링을 해보고 싶었는데 이번 배포에서는 보류할게요 ..

4. 그리고 이제와서 생각난 관련 문제는 ..ㅎㅎ 다른 컴포넌트(예를 들면 모달)들이 같은 포털을 사용하게 된다는 점인데, 개선한다면
  - 포털을 사용하는 컴포넌트가 더 생긴다면 각각의 portal root id를 만들고
  -  Portal에서 고정된 요소를 createPortal 하는게 아니라 props로 전달받은 id로 요소를 탐색하도록 할 수 있을 것 같아요.

iyu88 commented:
간밤에 논의했던 내용을 잠깐 정리하면
제가 생각했던 해결 방식은 다음과 같이 크게 세 가지 정도가 있을 것 같다고 말씀드렸었고,

> 1. Portal 마다 id 명시적으로 받게 하는 방법
> 2. Ref 를 root로 사용하도록  받는 방법
> 3. Toast 그대로 두고 Portal로 감싼 Toast만 가져다가 사용하게 하는 방법

그 중에서 3번에 대해서는 

> 그래서 3번 안을 생각하게 된게 일단 Docs에서 보여주고 싶은 내용은 Toast 그 자체에 대한 부분이니까
실제로 사용자가 가져다가 사용하는 Toast 는 한번 Portal 로 Wrapping된 녀석만 사용할 수 있게 하면
스토리에서 UI는 UI대로 확인하고
사용할 때는 Portal과 함께 사용할 수 있을 거라고 생각했어요

라고 말씀을 드렸어요.

우리 모두 스토리를 위해서 컴포넌트를 수정하는 것이 약간 마음에 걸리긴 하지만!

각각의 스토리를 정상적으로 보여주기 위해서 Portal은 제거하는 방향으로 가고 

Toast를 사용하는 입장을 고려한 **Portal로 감싼 Toast를 제공**하는 것이 좋겠다는 쪽으로 논의된 것 같아요.

배포하기 전에 해당 컴포넌트 작업이 필요하겠네요.

잘못된 부분은 댓글로 또 남겨주세요~

se030 commented:
스토리 문서에서 용례를 어떻게 보여줘야한다고 생각했는지에 차이가 있었던 것 같아요. 위 댓글을 쓸 때는 실제 쓰임에 따라서 화면 기준으로 보여져야 한다고 생각했는데, 현빈님 의견을 들어보니 스토리 문서 통일성을 위해서 영역 내부에서 보여질 수 있게 하는 것도 좋겠다고 생각했어요.

어쨌든 정리했던 결론을 생각해보면 이런저런 이슈가 있지만 Toast 컴포넌트 기능을 위해 필요한 Portal을 제거하는 것보다 스토리 용으로 사용할 수 있는 UI 부분을 별도 컴포넌트(네이밍은 생각이 안나요 ToastUI 뭐 .. 이런거)로 분리하는게 좋아보여요 ~

iyu88 commented:
> Portal을 제거하는 것보다 스토리 용으로 사용할 수 있는 UI 부분을 별도 컴포넌트로 분리하는게 좋아보여요 ~

Portal을 제거하지 않으면 스토리 용으로 만들 수 없지 않나? 라고 생각했는데, 왜냐하면 저는 

1. Toast에서 Portal 제거 
2. 현재 Toast를 Portal 로 감싼 ToastPortal (가명) 컴포넌트 생성 

라는 순서로 생각했어요.

곰곰이 생각해보니까 세영님은 

1. Toast에서 Portal 유지
2. 내부 내용만 별도로 ToastUI(가명) 컴포넌트로 분리 

이렇게 생각하신게 맞나요??

se030 commented:
> > Portal을 제거하는 것보다 스토리 용으로 사용할 수 있는 UI 부분을 별도 컴포넌트로 분리하는게 좋아보여요 ~
> 
> Portal을 제거하지 않으면 스토리 용으로 만들 수 없지 않나? 라고 생각했는데, 왜냐하면 저는
> 
> 1. Toast에서 Portal 제거
> 2. 현재 Toast를 Portal 로 감싼 ToastPortal (가명) 컴포넌트 생성
> 
> 라는 순서로 생각했어요.
> 
> 곰곰이 생각해보니까 세영님은
> 
> 1. Toast에서 Portal 유지
> 2. 내부 내용만 별도로 ToastUI(가명) 컴포넌트로 분리
> 
> 이렇게 생각하신게 맞나요??

네 맞아요. 저는 라이브러리에서 제공하는 컴포넌트 Toast에는 기능이 포함되어 있어야 한다고 생각했어요.
스토리북을 작성하는 입장이 아닌 사용자 입장에서는 Portal 없이 토스트를 사용할 일이 없지 않을까요 ??

iyu88 commented:
맞아요. 

지금 같은 얘기하고 있는데 작업 방식만 다른 것 같아요.

지금 이 PR에서 작업을 진행할까요?

맞다면 컴포넌트 이름 좀 추천해주세요 ㅎㅎㅎ

se030 commented:
> 맞아요.
> 
> 지금 같은 얘기하고 있는데 작업 방식만 다른 것 같아요.
> 
> 지금 이 PR에서 작업을 진행할까요?
> 
> 맞다면 컴포넌트 이름 좀 추천해주세요 ㅎㅎㅎ

네 여기서 작업하면 좋을 것 같아요! 작명 기회는 ... 현빈님께 양보할게요 ㅋㅋ

iyu88 committed:
refactor: ToastOrigin 생성

Toast에서 Portal을 사용하고 있는 점이 스토리 렌더링에 부적절하다고 판단하여 ToastOrigin을 생성하고 Portal이 없는 컴포넌트를 제공합니다.

iyu88 committed:
fix: ToastIcon 타입 경로 수정

se030 commented on src/components/Toast/index.tsx:
```diff
@@ -1,158 +1,13 @@
-      clearTimeout(timer);
-    };
-  }, [open]);
+import ToastOrigin, { ToastProps } from './ToastOrigin';
```
ㅋㅋㅋㅋ 왠지 브라우저 origin이 생각나서 이런건 어떨까 와르르 생각해봤어요. 근데 좋은진 모르겠어요 .. 더 생각해볼게요
- ToastWithoutPortal
- ToastView
- SToast
- StyledToast

iyu88 replied:
쟁쟁한 후보들 중에서 ToastCore로 최종 결정되었습니다.

se030 commented on src/components/Toast/index.tsx:
```diff
@@ -1,158 +1,13 @@
 
+const Toast = (props: ToastProps) => {
   return (
     <Portal>
```
포털 사용 관련해서 토스트 쪽에서 수정했으면 하는 부분이 있는데요.

1. Portal id 필수 props로 변경
2. 기존 root id 상수를 Toast 컴포넌트 포털용 id로 수정
3. CdsProvider, Toast 컴포넌트에 적용

세가지는 배포 전에 작업 가능하면 좋을 것 같아요. 연관 작업이 아니라고 생각되면 이슈 생성해서 반영해주세요 ~

iyu88 replied:
전부 반영했습니다 👍

iyu88 committed:
refactor: Portal Id 변경

각 컴포넌트마다 다른 root를 사용하기 위해 Portal이 Id를 필수적으로 받도록 변경합니다.
PORTAL_TOAST_ROOT_ID 를 새롭게 선언합니다.

iyu88 committed:
refactor: ToastCore 이름 변경

ToastOrigin 이름을 ToastCore로 변경합니다.
Portal에 Toast를 위한 root id를 명시합니다.
